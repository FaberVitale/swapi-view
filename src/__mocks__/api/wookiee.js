export default {
  whrascwo: "Yrahoahwh IV",
  rcooaoraaoahoowh_akworcahoowa: "24",
  oorcrhahaoraan_akworcahoowa: "4818",
  waahrascwoaoworc: "10200",
  oaanahscraaowo: "aowoscakworcraaowo, aorcooakahoaraan",
  rrrcrahoahaoro: "1 caorawhwararcwa",
  aoworcrcraahwh: "shhuwhrranwo, rcraahwhwwoorcwocaoc",
  churcwwraoawo_ohraaoworc: "8",
  akooakhuanraaoahoowh: "1000",
  rcwocahwawowhaoc: [],
  wwahanscc: ["acaoaoakc://cohraakah.oaoo/raakah/wwahanscc/1/"],
  oarcworaaowowa: "2014-12-10T11:37:19.144000Z",
  wowaahaowowa: "2014-12-20T20:58:18.421000Z",
  hurcan: "acaoaoakc://cohraakah.oaoo/raakah/akanrawhwoaoc/3/"
};
