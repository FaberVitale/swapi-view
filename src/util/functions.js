//@flow
import _get from "lodash.get";

export const getCurried = (keys: Array<string>, or: any) => (val: Object) =>
  _get(val, keys, or);

export const identity = (id: mixed) => id;

export const notEqual = (a: mixed) => (b: mixed) => a !== b;

export const equals = (a: mixed) => (b: mixed) => a === b;

export const isTruthy = (val: mixed) => !!val;

export const equalsTo = (...args: Array<mixed>) => (val: mixed) =>
  args.some(equals(val));

export const getDisplayName = (hocName: string, Comp: any) =>
  `with${hocName}(${Comp.displayName || Comp.name || "Component"})`;

export const warn: (...msgs: Array<any>) => void =
  process.env.NODE_ENV !== "production" &&
  window.console &&
  typeof window.console.warn === "function"
    ? window.console.warn
    : () => {};
