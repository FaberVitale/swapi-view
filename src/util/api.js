//@flow
import { warn } from "./functions";

/* Thrown when a GET request returns with a status !== 200.
 * based on https://stackoverflow.com/a/5251506
 */
export function ApiError(url: string, status: number) {
  this.url = url;
  this.status = status;
  this.name = "ApiError";
  this.message = `url: ${url}, status: ${status}`;
  this.stack = new Error().stack;
}
// $FlowFixMe
ApiError.prototype = new Error();

/* higher order function that given an optional timeout
  returns a fetch-like function that
  on fulfill will returns the response,
  on rejection returns the following errors:
  
  1 - TypeError: Network error or Permission error
  2 - ApiError: If the response status isnt 200
  3 - SyntaxError: Invalid JSON
  4 - an Error with message "timeout" if the request is hanging
  5 - RangeError - If the response status isnt in range [200; 599]

  spec: https://fetch.spec.whatwg.org/
*/
export const fetchData = (timeout?: number = 10000) => {
  let wait: number;

  if (!Number.isSafeInteger(timeout) || timeout < 500) {
    warn(
      `timeout is not a valid number:
       expected an Integer in range [500, 2^31 -1], got: ${timeout}
       timeout set to 10s`
    );
    wait = 10000;
  }
  wait = timeout;

  return async (url: string | URL, init?: RequestOptions) => {
    let res;

    res = await new Promise((resolve, reject) => {
      let tRef = setTimeout(() => {
        reject(new Error("timeout"));
      }, wait);

      const req = init ? fetch(url, init) : fetch(url);

      req.then(response => {
        clearTimeout(tRef);
        resolve(response);
      }, reject);
    });

    if (res.status === 200) {
      return res;
    }

    throw new ApiError(
      typeof url !== "string" ? url.toString() : url,
      res.status
    );
  };
};
