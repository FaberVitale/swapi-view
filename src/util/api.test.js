import * as api from "./api";
import { identity } from "./functions";

describe("util/api", () => {
  describe("ApiError", () => {
    const err = new api.ApiError("test.test", "500");

    const withoutStack = Object.assign({}, err);

    Object.setPrototypeOf(withoutStack, api.ApiError.prototype);

    test("ApiError", () => {
      expect(err).toBeInstanceOf(Error);
      expect(withoutStack).toMatchSnapshot();
      expect(typeof err.stack).toBe("string");
    });
  });

  describe("fetchData", () => {
    jest.useFakeTimers();

    let res = { json: () => obj, status: 200, statusText: "200" };

    const url = "test.test";
    const reqInfo = { method: "GET", mode: "same-origin" };

    const obj = { succcess: true };

    const resolvedFetch = () => Promise.resolve(res);
    const timeoutFetch = (...args) =>
      new Promise((resolve, reject) => {
        setTimeout(reject, 20000);
      });

    const apiErrorFetch = (...args) =>
      Promise.resolve(
        Object.assign({}, res, { status: 500, statusText: "500" })
      );

    let fetchMock = jest.fn();

    let realFetch;

    beforeAll(() => {
      if (window.fetch) {
        realFetch = window.fetch;
      }
      window.fetch = fetchMock;
    });

    afterAll(() => {
      if (realFetch) {
        window.fetch = realFetch;
      }
    });

    afterEach(() => {
      fetchMock.mockClear();
    });

    const timeoutAssertions = (promise, done) =>
      promise.then(identity, identity).then(maybeError => {
        expect(maybeError).toBeInstanceOf(Error);
        expect(fetchMock.mock.calls[0][0]).toBe(url);
        expect(maybeError.message).toBe("timeout");
        done();
      });

    it("returns the response on status === 200", done => {
      fetchMock.mockImplementation(resolvedFetch);

      api
        .fetchData()(url)
        .then(retrieved => {
          expect(retrieved).toBe(res);
          expect(fetchMock.mock.calls).toHaveLength(1);
          done();
        });
    });

    it("it rejects if status isnt 200", done => {
      fetchMock.mockImplementation(apiErrorFetch);

      api
        .fetchData()(url)
        .then(identity, identity)
        .then(maybeError => {
          expect(maybeError).toBeInstanceOf(api.ApiError);
          expect(fetchMock.mock.calls[0][0]).toBe(url);
          done();
        });
    });

    it("returns an error with message timeout if the request takes too long", done => {
      const timeout = 10000;

      fetchMock.mockImplementation(timeoutFetch);

      timeoutAssertions(api.fetchData(timeout)(url), done);

      jest.runAllTimers();
    });

    it("if no argument is provided sets timeout to 10s", done => {
      fetchMock.mockImplementation(timeoutFetch);

      timeoutAssertions(api.fetchData()(url), done);

      jest.runTimersToTime(10001);
    });

    it("sets timeout to 10s if an invalid integer is provided", done => {
      fetchMock.mockImplementation(timeoutFetch);

      timeoutAssertions(api.fetchData(-2)(url), done);

      jest.runTimersToTime(10001);
    });

    it("if a requestInfo is passed it is used to perform a fetchRequest", done => {
      fetchMock.mockImplementation(resolvedFetch);

      api
        .fetchData()(url, reqInfo)
        .then(retrieved => {
          expect(retrieved).toBe(res);
          expect(fetchMock.mock.calls).toHaveLength(1);
          expect(fetchMock.mock.calls[0]).toEqual([url, reqInfo]);
          done();
        });
    });
  });
});
