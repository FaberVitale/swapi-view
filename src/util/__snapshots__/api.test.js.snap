// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util/api ApiError ApiError 1`] = `[ApiError: url: test.test, status: 500]`;
